### PR TITLE
feat: add prefer-user-facing-locators rule

### PR DESCRIPTION
**Pull request by Haberkamp**
This merge request adds a rule that prefers the usage of user-facing locators over `page.locator()`.  Using user-facing selectors makes the tests more robust.

Resolves: #123

### CODE REVIEW

**Comment by mskelton** [comment]
Not really a fan of the name here, what about

- `prefer-testing-library-methods`
- `disallow-raw-locators`
- ...

Honestly not sure I like those options either, hmm.

**Comment by grant** [comment]
Would definitely use this rule! the `prefer-` prefix makes sense to me.

**Comment by mskelton** [comment]
Problem with `disallow-page-locator-selector` is that there is also `frame.locator`, `locator.locator`, etc.

I think I would like to proceed with `disallow-raw-locators`, it makes the most sense of all the options given.

**Comment by Haberkamp** [comment]
@mskelton That's a good point you're making. I would then change the name of the rule to your suggested `disallow-raw-locators`. And to be sure that I'm understanding you correctly. Should I also modify the rule in a way that will disallow the usage of `frame.locator` and `locator.locator`?

**Comment by mskelton** [comment]
Yes, the rule should detect all usages of `.locator`, not just `page.locator`.

**Comment by Haberkamp** [comment]
Talking about the naming again. I think it would make sense to align it with other rules like `no-page-pause` or `no-useless-await`. So the rule would be named `no-raw-selector`.

Another thing: should this rule be included in the recommended config and with which level of severity should this rule report?

**Comment by mskelton** [comment]
Yeah, good call, `no-raw-selector` is better.

This should definitely NOT be in the recommended config. While the other methods are better when possible to use, there are still plenty of valid use cases for `.locator()` and I do not want that warning users all the time unnecessarily.

**Comment by Haberkamp** [comment]
Ok, I've made the changes you've requested.

File: docs/rules/prefer-user-facing-locators.md
Comment (mskelton) [inline docs/rules/prefer-user-facing-locators.md:1]: This file needs to be renamed to `no-raw-selectors`

File: docs/rules/prefer-user-facing-locators.md [hunk 1]
```diff
@@ -0,0 +1,26 @@
+## Disallow using raw selectors (`no-raw-selector`)
```
Comment (mskelton) [inline docs/rules/prefer-user-facing-locators.md:1]: Wait a minute, I thought we agreed on `no-raw-locators`?

File: docs/rules/prefer-user-facing-locators.md
Comment (Haberkamp) [reply docs/rules/prefer-user-facing-locators.md:1]: The name of the file is now `no-raw-locators.md`. I've aligned the name with your suggestion from the other conversation.

File: test/spec/no-raw-locators.spec.ts [hunk 2]
```diff
@@ -0,0 +1,65 @@
+import rule from '../../src/rules/no-raw-locators';
+import { runRuleTester, test } from '../utils/rule-tester';
+
+const messageId = 'noRawSelector';
```
Comment (mskelton) [inline test/spec/no-raw-locators.spec.ts:4]:
```suggestion
const messageId = 'noRawLocator';
```

**Review by mskelton — APPROVED** [review]

**Comment by seanlaff** [comment]
Should this rule be added to the list of rules in the readme?

**Comment by mskelton** [comment]
@seanlaff Indeed it should, I missed that before. If you are willing to open a PR, that'd be appreciated!

**Comment by mskelton** [comment]
@seanlaff It's in the readme now

**Pull request merged**